src/Compilers/CSharp/Portable/Symbols/Source/QuickAttributeChecker.cs

```diff
@@ -0,0 +1,166 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    /// <summary>
+    /// The QuickAttributeChecker applies a simple fast heuristic for determining probable
+    /// attributes of certain kinds without binding attribute types, just by looking at the final syntax of an
+    /// attribute usage.
+    /// </summary>
+    /// <remarks>
+    /// It works by maintaining a dictionary of all possible simple names that might map to the given
+    /// attribute.
+    /// </remarks>
+    internal sealed class QuickAttributeChecker
+    {
+        private readonly Dictionary<string, QuickAttributes> _nameToAttributeMap;
+        private static QuickAttributeChecker _lazyQuickAttributeChecker;
+
+#if DEBUG
+        private bool _sealed;
+#endif
+
+        internal static QuickAttributeChecker Predefined
+        {
+            get
+            {
+                if (_lazyQuickAttributeChecker is null)
+                {
+                    Interlocked.CompareExchange(ref _lazyQuickAttributeChecker, CreateQuickAttributeChecker(), null);
+                }
+
+                return _lazyQuickAttributeChecker;
+            }
+        }
+
+        private static QuickAttributeChecker CreateQuickAttributeChecker()
+        {
+            var result = new QuickAttributeChecker();
+            result.AddName(AttributeDescription.TypeIdentifierAttribute.Name, QuickAttributes.TypeIdentifier);
+            result.AddName(AttributeDescription.TypeForwardedToAttribute.Name, QuickAttributes.TypeForwardedTo);
+
+#if DEBUG
+            result._sealed = true;
+#endif
+            return result;
+        }
+
+        private QuickAttributeChecker()
+        {
+            _nameToAttributeMap = new Dictionary<string, QuickAttributes>();
+            // NOTE: caller must seal
+        }
+
+        private QuickAttributeChecker(QuickAttributeChecker previous)
+        {
+            _nameToAttributeMap = new Dictionary<string, QuickAttributes>(previous._nameToAttributeMap);
+            // NOTE: caller must seal
+        }
+
+        private void AddName(string name, QuickAttributes newAttributes)
+        {
+#if DEBUG
+            Debug.Assert(!_sealed);
+#endif
+            var currentValue = QuickAttributes.None;
+            _nameToAttributeMap.TryGetValue(name, out currentValue);
+
+            QuickAttributes newValue = newAttributes | currentValue;
+            _nameToAttributeMap[name] = newValue;
+
+            // We allow "Name" to bind to "NameAttribute"
+            if (name.EndsWith("Attribute", StringComparison.OrdinalIgnoreCase))
+            {
+                _nameToAttributeMap[name.Substring(0, name.Length - "Attribute".Length)] = newValue;
```

Comment:
>_nameToAttributeMap[name.Substring(0, name.Length - "Attribute".Length)] = newValue; [](start = 16, length = 84)

This looks fragile. Could we drop some bits from the value? I think we should simply combine newAttributes with whatever with already have for this key, similar as we do above.